examples/autoencoders/ae/models/modules.py

```diff
@@ -0,0 +1,696 @@
+        h_ = ops.reshape(h_, (b, c, h, w))
+
+        h_ = self.proj_out(h_)
+
+        return x + h_
+
+
+class AttnBlock3D(nn.Cell):
+    def __init__(self, in_channels, dtype=ms.float32):
+        super().__init__()
+        self.in_channels = in_channels
+        self.dtype = dtype
+
+        self.bmm = ops.BatchMatMul()
+        self.norm = Normalize(in_channels)
+
+        # TODO: 1x1 conv3d can be replaced with Linear
+        self.q = CausalConv3d(in_channels, in_channels, kernel_size=1, stride=1).to_float(dtype)
+        self.k = CausalConv3d(in_channels, in_channels, kernel_size=1, stride=1).to_float(dtype)
+        self.v = CausalConv3d(in_channels, in_channels, kernel_size=1, stride=1).to_float(dtype)
+        self.proj_out = CausalConv3d(in_channels, in_channels, kernel_size=1, stride=1).to_float(dtype)
+
+    def construct(self, x):
+        h_ = x
+        h_ = self.norm(h_)
+        q = self.q(h_)
+        k = self.k(h_)
+        v = self.v(h_)
+
+        # compute attention
+        b, c, t, h, w = q.shape
+        q = ops.reshape(q, (b, c, h * w))
+        q = ops.transpose(q, (0, 2, 1))  # b,hw,c
+        k = ops.reshape(k, (b, c, h * w))  # b,c,hw
+        w_ = self.bmm(q, k)  # b,hw,hw    w[b,i,j]=sum_c q[b,i,c]k[b,c,j]
+
+        w_ = w_ * (int(c) ** (-0.5))
+        w_ = ops.Softmax(axis=2)(w_)
+
+        # attend to values
+        v = ops.reshape(v, (b, c, h * w))
+        w_ = ops.transpose(w_, (0, 2, 1))  # b,hw,hw (first hw of k, second of q)
+        h_ = self.bmm(v, w_)  # b, c,hw (hw of q) h_[b,c,j] = sum_i v[b,c,i] w_[b,i,j]
+        h_ = ops.reshape(h_, (b, c, h, w))
+
+        h_ = self.proj_out(h_)
+
+        return x + h_
+
+
+def make_attn(in_channels, attn_type="vanilla", dtype=ms.float32):
+    assert attn_type in ["vanilla", "vanilla3D"], f"attn_type {attn_type} not supported"
+    _logger.debug(f"making attention of type '{attn_type}' with {in_channels} in_channels")
+    if attn_type == "vanilla":
+        return AttnBlock(in_channels, dtype=dtype)
+    elif attn_type == "vanilla3D":
+        return AttnBlock3D(in_channels, dtype=dtype)
+
+
+# used in vae
+class Encoder(nn.Cell):
+    def __init__(
+        self,
+        *,
+        ch,
+        out_ch,
+        ch_mult=(1, 2, 4, 8),
+        num_res_blocks,
+        attn_resolutions,
+        dropout=0.0,
+        resamp_with_conv=True,
+        in_channels,
+        resolution,
+        z_channels,
+        double_z=True,
+        use_linear_attn=False,
+        attn_type="vanilla",
+        dtype=ms.float32,
+        upcast_sigmoid=False,
+        **ignore_kwargs,
+    ):
+        super().__init__()
+        # if use_linear_attn: attn_type = "linear"
+        self.ch = ch
+        self.temb_ch = 0
+        self.num_resolutions = len(ch_mult)
+        self.num_res_blocks = num_res_blocks
+        self.resolution = resolution
+        self.in_channels = in_channels
+        self.dtype = dtype
```

Comment:
I think adding an explicit `dtype` to each cell unnecessarily increases complexity. I believe a better way is to control types through AMP.